sequencer/batchbuilder.go

```diff
@@ -44,30 +45,57 @@ func (s *Sequencer) tryToProcessTx(ctx context.Context, ticker *time.Ticker) {
 	}
 
 	// Get next tx from the pool
-	log.Info("getting pending tx from the pool")
-	tx, err := s.pool.GetTopPendingTxByProfitabilityAndZkCounters(ctx, s.calculateZkCounters())
-	if err == pgpoolstorage.ErrNotFound {
-		log.Infof("there is no suitable pending tx in the pool, waiting...")
-		waitTick(ctx, ticker)
-		return
-	} else if err != nil {
-		log.Errorf("failed to get pending tx, err: %v", err)
-		return
+	log.Info("getting pending txs from the pool")
+	startBuildBatchTime := time.Now()
+	for !s.isZkCountersMoreThanMax(s.sumZkCounters) {
+		if startBuildBatchTime.Add(s.cfg.MaxWaitTimeForSuitableTxToAppear.Duration).Before(time.Now()) && len(s.pendingTxs) > 0 {
+			log.Info("time to gather txs are passed, gathering ended")
+			break
+		}
+
+		pendTx, err := s.pool.GetTopPendingTxByProfitabilityAndZkCounters(ctx, s.maxZkCountersSubPassedZkCounters(s.sumZkCounters), s.pendingTxsHashes)
+		if err == pgpoolstorage.ErrNotFound {
+			log.Infof("there is no suitable pending tx in the pool, waiting...")
+			waitTick(ctx, ticker)
+			continue
+		} else if err != nil {
+			log.Errorf("failed to get pending tx, err: %v", err)
+			return
+		} else {
+			log.Infof("adding pending txs to pending tx array, hash: %s", pendTx.Hash().String())
+		}
```

Comment:
This last `else` statement is unnecessary, you can just put the log on top of `SumUpZkCounters`